chisel/src/config.rs

```diff
@@ -0,0 +1,570 @@
+//! Configuration utilities for chisel modules.
+//! The two main traits are FromYaml and FromArgs, for config-driven and oneliner mode
+//! respectively.
+//! This module implements the ChiselConfig structure, which contains the information for a number
+//! of chisel executions.
+//! A ChiselConfig contains a list of ruleset configurations and their names.
+//! A ruleset configuration contains a set of top-level options and a set of module configurations to execute.
+//! The aforementioned traits are used to generate ChiselConfigs.
+
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::fmt::{self, Display};
+
+use serde_yaml::Value;
+
+/// Main trait for generating a configuration usable by the driver from deserialized YAML.
+pub trait FromYaml {
+    // yaml: top-level yaml element generated by parser
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+/// Main trait for generating a configuration usable by the driver from CLI arguments.
+pub trait FromArgs {
+    // Supported formatting example:
+    // modules: "remapimports, trimexports, verifyimports"
+    // options: "remapimports.preset=ewasm, trimexport.option=foo, verifyimports.a=b"
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+/// A list of rulesets for a single chisel run.
+/// Left-hand value: name of ruleset
+/// Right-hand value: ruleset configuration
+#[derive(Debug)]
+pub struct ChiselConfig(VecDeque<(String, Ruleset)>);
+
+/// A single ruleset, containing options and a set of module configurations.
+#[derive(Debug)]
+pub struct Ruleset {
+    /// Top-level ruleset configurations, such as the input/output file paths.
+    opts: HashMap<String, String>,
```

Comment:
Nitpick: some parts use `opt`/`opts` other use `option`/`options`. I'd just go with the longer for readablity.

Reply:
This applies throughout the file :)